Refactor power distribution Result

daniel-zullo-frequenz opened the pull request:
Since Frequenz-SDK is already using Python 3.11 as a minimum version, Result can now be defined as a union type of the different results instead of using inheritance.

daniel-zullo-frequenz committed:
Refactor power distribution Result

Since Frequenz-SDK is already using Python 3.11 as a minimum version,
Result can now be defined as a union type of the different results
instead of using inheritance.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax reviewed (commented):
Amazing. Even when this should be probably mainly backwards compatible, I would mention it in the release notes and recommend using `match` to figure out what kind of a result a `Result` is. It would be also good to look for both internal code using `Result`s and examples and replace those with `match`.

daniel-zullo-frequenz commented on src/frequenz/sdk/actor/power_distributing/result.py:
```diff
@@ -48,19 +36,19 @@ class _BaseSuccessMixin:
-# otherwise the request attribute will be last in the dataclass constructor
-# because of how MRO works.
+# We need to put the _BaseSuccessMixin before _BaseResultMixin in the
+# inheritance list to make sure that the Result attributes appear before the
```
```suggestion
# inheritance list to make sure that the _BaseResultMixin attributes appear before the
```

daniel-zullo-frequenz commented:
> I would mention it in the release notes and recommend using match to figure out what kind of a result a Result is. It would be also good to look for both internal code using Results and examples and replace those with match.

I added an example as I only found assertions in tests to check only one type of results at a time where `match` was not suitable. Also added the example in the release notes, please let me know if that's what you had in mind

llucax commented on RELEASE_NOTES.md:
````diff
@@ -10,6 +10,51 @@
+        match result:
+            case Success() as success:
+                print(f"Power request was successful: {success}")
+            case PartialFailure() as partial_failure:
+                print(f"Power request was partially successful: {partial_failure}")
+            case OutOfBounds() as out_of_bounds:
+                print(f"Power request was out of bounds: {out_of_bounds}")
+            case Error() as error:
+                print(f"Power request failed: {error}")
+            case _:
+                assert False, "Should not reach this point"
+
+    request = Request(
+        namespace="TestChannel",
+        power=Power.from_watts(123.4),
+        batteries={8, 18},
+    )
+
+    results: list[Result] = [
+        Success(
+            request,
+            succeeded_power=Power.from_watts(123.4),
+            succeeded_batteries={8, 18},
+            excess_power=Power.zero(),
+        ),
+        PartialFailure(
+            request,
+            succeeded_power=Power.from_watts(103.4),
+            succeeded_batteries={8},
+            excess_power=Power.zero(),
+            failed_batteries={18},
+            failed_power=Power.from_watts(20.0),
+        ),
+        OutOfBounds(request, bounds=PowerBounds(0, 0, 0, 800)),
+        Error(request, msg="The batteries are not available"),
+    ]
+
+    for r in results:
+        handle_power_request_result(r)
+  ```
````
I think this might be too much for a release note, I think this should suffice:

~~~suggestion
  ```python
  from typing import assert_never
  result: Result = some_operation()
  match result:
      case Success() as success:
          print(f"Power request was successful: {success}")
      case PartialFailure() as partial_failure:
          print(f"Power request was partially successful: {partial_failure}")
      case OutOfBounds() as out_of_bounds:
          print(f"Power request was out of bounds: {out_of_bounds}")
      case Error() as error:
          print(f"Power request failed: {error}")
      case _ as unreachable:
          assert_never(unreachable)
  ```
~~~

Using `assert_never` makes sure that if a new result is added, the `match` is forced to handle it (otherwise it will be a `mypy` error).

daniel-zullo-frequenz replied:
right, that was indeed too much :angel: Thanks for the suggestions!

llucax commented on src/frequenz/sdk/actor/power_distributing/result.py:
````diff
@@ -109,3 +97,64 @@ class OutOfBounds(Result):
+Example: Handling power distribution results
+
+    ```python
+    from frequenz.sdk.actor.power_distributing import (
````
```suggestion
    from typing import assert_never

    from frequenz.sdk.actor.power_distributing import (
```

llucax commented on src/frequenz/sdk/actor/power_distributing/result.py:
```diff
@@ -109,3 +97,64 @@ class OutOfBounds(Result):
+                print(f"Power request was out of bounds: {out_of_bounds}")
+            case Error() as error:
+                print(f"Power request failed: {error}")
+            case _:
+                assert False, "Should not reach this point"
```
```suggestion
            case _ as unreachable:
                assert_never(unreachable)
```
